lib/src/command/run.dart

```diff
@@ -86,4 +91,98 @@ class RunCommand extends PubCommand {
     });
     await flushThenExit(exitCode);
   }
+
+  /// Implement a v2 mode for use in `dartdev run`.
+  ///
+  /// Usage: `dartdev run [package[:command]]`
+  ///
+  /// If `package` is not given, defaults to current root package.
+  /// If `command` is not given, defaults to name of `package`.
+  /// If neither `package` or `command` is given and `command` with name of
+  /// the current package doesn't exist we fallback to `'main'`.
```

Comment:
Why do we need the magic fallback to `'main'`? Do we have templates that add a `bin/main.dart` by default?

I worry that adding something new - `bin/<package>.dart` changes the behavior of `pub run`. So if I've had a `bin/main.dart` and have been happily working, adding a new file gives new different magic behavior...

What does the error message look like if you `pub run` and there is no `bin/<package.dart` or `bin/main.dart`? Does it complain about missing `main.dart` only? That could encourage more people to add that, and if people publish those would the workflow end up looking like `pub run some_package:main` for most use cases? Or does `pub run some_package` also run` bin/main.dart` in that package?

Reply:
It only does the fallback if `bin/main.dart` exists. @mit-mit  wanted it because he wanted `dartdev run` without any arguments to work with out existing templates. I see that as valuable, but on the other hand I agree that using `bin/main.dart` will give a bad workflow in other cases.

I think the reasoning was that if you make a console app that is not distributed on `pub.dev`, using `bin/main.dart` is what our templates have done for years, and the drawbacks to supporting that are fairly slim.
IMO, people aren't going to notice that `bin/main.dart` works, because it's only a fallback that is used when:
 * calling without arguments, and,
 * `bin/<packageName>.dart` does exist, and,
 * `bin/main.dart` does exist.

Reply:
I think as long as the error message if none off the fallbacks exist mentions `bin/<packageName>.dart`, and does _not_ mention `bin/main.dart` I think it's probably OK.

I hope to avoid new things that would subtly encourage that pattern, but I'm fine with attempting to keep compatibility with the other places that were already encouraging that pattern.

Reply:
What advantages does `bin/<packageName>.dart` have over `bin/main.dart`?

Flutter apps seem to use `lib/main.dart`, hmm...

Reply:
@jonasfj you meant "bin/<packageName>.dart does NOT exist, and" right?

Reply:
> What advantages does `bin/<packageName>.dart` have over `bin/main.dart`?

It's what runs when you do `pub run <packageName>`, `pub run test` runs `bin/test.dart` from the test package. When publishing a package that is meant to be run this way this is the naming scheme that should be used.

> Flutter apps seem to use `lib/main.dart`, hmm...

Flutter ignored our conventions before we had an opportunity to add in hints and stuff and now we are stuck with it. FWIW `lib/main.dart` will almost always be flutter, which means it will almost never work with `pub run`.

Reply:
> I hope to avoid new things that would subtly encourage that pattern, but I'm fine with attempting to keep compatibility with the other places that were already encouraging that pattern.

I feel exactly the same way. Which is also why `bin/main.dart` will not be used if you do `dartdev run <packageName>` -- it only works for the local root package, ie. `dartdev run` (without arguments).

Reply:
I'm open towards changing what the recommended file name is for the entrypoint; we can always change the template if needed.

But remember that we have two different types of packages -- library packages (which are used by apps), and app packages (the CLI, web, mobile app being built). Do we think `bin/<packageName>.dart` is a better choice for apps?

Reply:
> Do we think `bin/<packageName>.dart` is a better choice for apps?

It's the current convention for apps, example: [mono_repo](https://github.com/dart-lang/mono_repo/tree/master/mono_repo/bin), [build_runner](https://github.com/dart-lang/build/tree/master/build_runner/bin), [test](https://github.com/dart-lang/test/tree/master/pkgs/test/bin), [webdev](https://github.com/dart-lang/webdev/tree/master/webdev/bin), [dartdoc](https://github.com/dart-lang/dartdoc/tree/master/bin), [grinder](https://github.com/google/grinder.dart/tree/master/bin),  [tota](https://github.com/tota-dart/tota/tree/master/bin), [dart_up](https://github.com/regiostech/dart_up/tree/master/bin), [lefthook](https://github.com/drt-pkgz/lefthook/tree/master/bin) -- this was just looking at package that use `cli_util`.

The only counter example I found was [fvm](https://github.com/leoafarias/fvm/tree/master/bin), which used `bin/main.dart`.

Reply:
I guess my only concern is that I don't understand *why* `<packageName>.dart` was originally selected? Using `main` as the main entry point is pretty common across a range of languages. Does it have some functional benefit that I'm not seeing?

Reply:
I never liked/understood that convention either.
Neither for the main library (`lib/<packageName>.dart`) nor binary (`bin/packageName>.dart`).
There is too much repetition.

Reply:
Python has `__init__.py`, node has `index.js`, rust has `mod.rs`, I don't know that `<packageName>.dart` is such a bad choice.

Reply:
I guess this is not the right place to take this discussion.
...
But <packageName>.dart is bad (ok, let's say it has drawbacks) for at least the following reasons:

* if we rename the package we have to rename the main file, making this a more difficult operation (also making it even harder than it already is to vendor a package inside another - a use case I know you care about) 

* it made this feature (that just seems so obvious): https://github.com/dart-lang/language/issues/649 be non-trivial to specify.

I guess it comes down down to composability. - why would the name of the main library have to depend on the surrounding package?

Reply:
> Which is also why `bin/main.dart` will not be used if you do `dartdev run <packageName>` -- it only works for the local root package, ie. `dartdev run` (without arguments).

This is what I worry about, but maybe I'm imagining problems that don't exist.

What I'm worried is that `pub run` is conceptually a shorthand for "Run this package" becomes associated with `bin/main.dart` because that works. We know that `pub run <package>` is shorthand for "Run that named package". But now there is a difference in behavior between "Run this package" and "Run that package".


Regarding `lib/<package>.dart` - it's true that it's somewhat unnecessary repetition, but FWIW I think `import 'package:something/lib.dart';` would end up with complaints about weird conventions too.

I don't think at this point its worth the churn to try to change it.